src/upload/upload.handler.ts

```diff
@@ -0,0 +1,56 @@
+import fs from 'fs'
+import Libp2p from 'libp2p'
+import { subscribeForOffer } from '../communication'
+import { UploadJobStatus } from '../definitions'
+import { loggingFactory } from '../logger'
+import { ProviderManager } from '../providers'
+import UploadJob from './upload.model'
+
+const logger = loggingFactory('upload.handler')
+type UploadRouteHandler = (req: any, res: any) => Promise<void>
+
+export default function (storageProvider: ProviderManager, libp2p: Libp2p): UploadRouteHandler {
+  return async (req: any, res: any): Promise<void> => {
+    const { offerId, peerId, account } = req.body
+
+    if (!req.file) {
+      return res.status(422).json({
+        error: 'File needs to be provided.'
+      })
+    }
+
+    logger.info(`Receive file ${req.file.filename} for offer ${offerId}, peerId ${peerId}`)
+
+    // Create upload job
+    const job = await UploadJob.create({
+      offerId,
+      peerId,
+      account,
+      meta: { filename: req.file.originalname },
+      status: UploadJobStatus.UPLOADING
+    })
+    logger.info('Job created')
+
+    // Read file from fs and start uploading to storage
+    const data = await fs.promises.readFile(req.file.path)
```

Comment:
Uffff, if we are supporting up to 1 GB files I am not sure if we should not use streaming option. This will load the whole file into memory. If this get triggered 3-4 times at the same time the service will run of RAM...

Reply:
Yeah, lets use `ReadableStream` as described in [`js-ipfs` docs](https://github.com/ipfs/js-ipfs/blob/master/docs/core-api/FILES.md#ipfsadddata-options).

Reply:
I know you tried to pipe directly the upload stream to IPFS, that would be the best, but if you are not able to make that work, then we should at least use streams from reading from the tmp stored file for upload to IPFS...

Reply:
Done in https://github.com/rsksmart/rif-marketplace-upload-service/pull/14